assets/src/edit-story/utils/usePreventWindowUnload.js

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useEffect } from 'react';
+
+/**
+ * This control the listener register flow.
+ * It should be declared outside of the hook to avoid recreate different references of the registered handler in order to make posible remove it later from multiple contexts.
+ *
+ * @param {Object} event beforeunload Event object
+ */
+
+export const beforeUnloadListener = (event) => {
+  event.preventDefault();
+  event.returnValue = '';
+};
+
+/**
+ * Prevents window unloads without a confirmation prompt.
+ *
+ * @param {boolean} condition A condition that will control when register/unregister the listener to prevent window unload
+ */
+function usePreventWindowUnload(condition = false) {
```

Comment:
The way it's written, this is a rather dangerous hook to be run wherever. It should only be ever invoked in a minimal number of places (ideally only one place). One option is to make this code private to either `HistoryProvider`, or `useSaveStory`, or maybe `StoryProvider`. And "has changes" signal somehow calculated there or "saved" signal is propagated from `useSaveStory`. Generally, such a singleton task is better suited for a provider/private hook than a reusable hook.

Reply:
Initially, I created something under the HistoryProvider scope, but only for reusability of the logic, I moved to a hook since we will need to control this prompt in more places (For instance, uploads not completed), not only during the save. Anyway, I will take a look at your insights and move it back to some already existing stuff that controls the history and/or saves actions.

Reply:
We just need it to be extensible by e.g. uploads in progress and similar - is that still doable?

Reply:
Yup.

Reply:
Created a ticket for upload files - https://github.com/google/web-stories-wp/issues/971